### ISSUE TITLE
Enhance the WoT view

### ISSUE THREAD

**Issue by Insoleet** [ucoin-io/sakia#208]
I think we could enhance the wot view so that he could be animated, clickable, browsed...

This python library looks pretty good to do such a thing : 
https://www.nodebox.net/code/index.php/Graph
https://github.com/nodebox/nodebox-opengl

https://jakevdp.github.io/blog/2013/03/23/matplotlib-and-the-future-of-visualization-in-python/

https://github.com/bokeh/bokeh

**Comment by Insoleet** [comment]
I guess we could use this library finally : 
http://networkx.github.io/

It seems easy to use with PyQt, and would make the WoT view code much simpler : 
http://stackoverflow.com/questions/24833574/using-networkx-to-position-nodes-on-pyqt-qgraphicsscene

What do you think, @vtexier  ?

**Comment by vtexier** [comment]
Just before coding the actual WoT, I had found networkx, but wasn't able to use networkx layout in Qt.

The stackoverflow link seems to be the the missing link I was waiting for...

It definitely worth a try !

**Comment by Insoleet** [comment]
From what I can see with these first experimentations, networkx will be a nice tool to show big web of trusts. But the basic web of trust, showing certifiers and certified, seems not possible with network because : 
- We can't duplicate nodes : we will have to identify nodes based on "certifier-[pubkey]" and "certified-[pubkey]"
- This is still annoying to do...
- We can't choose which node to display left and which node to display right. Of course, we could just implement our own layout...

Basically, I think we will use networkx to show big web of trusts, with the "steps" parameter. But the tab showing our certifiers and certified should use the current code.

**Comment by Insoleet** [comment]
Ok I finally found a way.

We can use multiple layouts to display different information, like : 
- A layout for the certifiers (from left to right)
- A layout for the certified (from right to left)
- A layout for the WoT path (from top to bottom)

It's really easy to do. @vtexier , do you have time to checkout the changes in this branch and tell me if you think it's easier to read ?

I think it's still better to depend on an external library for graph computations. It will help us maintain the code and it should be easier to implement new graph views now.

**Comment by M5oul** [comment]
To add in documentation:
```bash
pip install networkx numpy
```

**Comment by vtexier** [comment]
I am on the branch.
I display the certifications tab, but canvas stay empty, even with good dates for the time period.
Is it that you want me to check or is it the code ?

**Comment by Insoleet** [comment]
The certifications tab is empty and should stay until I develop it ^^

I would appreciate if you could take a look at the new WoT view and mostly at the code ! The view shouldn't have changed a lot (I did my best to reproduce the view you developed).But the code is now using networkx and I'd like to know what you think about it !

**Comment by vtexier** [comment]
You have done a great job ! I struggled to find a difference between my layout and the new one...
It is the same as I can tell !
I thought during some time I was on a wrong branch. ;-)

You manage to use the networkx get_shortest_path function, but there is a bug and it is not working... I 'll take a look soon.
You did not use the layout possibility of networkx and I guess it is because information is better viewed in the simple homemade view...
Graph layout will be more useful in a explorer view...

**Comment by Insoleet** [comment]
Exactly, networkx native layouts were not good the information shown in this tab. Did you find the new code clean and easy to read ?

**Comment by Insoleet** [comment]
So with an explorer tab, I'm able to get some nice graphs using networkx and graphviz library : 
![wot1](https://cloud.githubusercontent.com/assets/1170293/11942914/40f9543e-a83a-11e5-944c-30e2b442b858.png)
The image was rendered using `twopi` program from graphviz.
The problem is, that we then depend on graphviz, which can be annoying. As a fallback mode, if twopi is not available, I'm planning to use some defaults layouts which networkx can use, like spring_layout : 
![wot2](https://cloud.githubusercontent.com/assets/1170293/11943068/539a2838-a83b-11e5-8931-386e750c9bf8.png)

What do you think about it ?

Also, we should probably not explore the whole WoT. A limit should be defined by the user, but this will come later...

**Comment by vtexier** [comment]
I can see that you managed to use network layout information and apply it in a qt canvas ! Wonderful !
To answer you on the code, yes the code is clean and readable for my opinion.

**Comment by c-geek** [comment]
Would it be possible to have colors for steps? This would imply to focus on a particular individual, and from him:

* we colorize 1st step links as green
* we colorize 2nd step links as yellow-green
* we colorize 3rd step links as yellow

Maybe this would be another view, not showing all of the links but just the few that reflects the WoT diameter from a particular member.

**Comment by Insoleet** [comment]
Yes this kind of feature is possible. I will try to implement it in the real explorer view cleanly using TDD, this was more of a POC !

**Comment by Insoleet** [comment]
Ok so I implemented almost everything needed. I'll need to refactor how we display nodes and arcs to be able to have multiple colors, styles, displays. You can still check out the current behaviour in the branch `feature/explorer_tab` !

What remains to be done : 
 - Add an animation on the current identity being looked up
 - Change colors of arcs and nodes depending on the distance to the current identity
![wot4](https://cloud.githubusercontent.com/assets/1170293/12081040/2b3fdc56-b26e-11e5-87d9-23db0658a038.png)

**Comment by c-geek** [comment]
Beautiful!

**Comment by M5oul** [comment]
How could I install pygraphviz, I have try with pip, but I get an error. It seems libcgraph dependency is missing.
```bash
(env-sakia) [moul@moulinette sakia]$ sudo pip install pygraphviz --upgrade
Mot de passe [sudo] de moul :
Collecting pygraphviz
  Using cached pygraphviz-1.3.1.tar.gz
Installing collected packages: pygraphviz
  Running setup.py install for pygraphviz
    Complete output from command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-lDPgM8/pygraphviz/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-Hcz4ba-record/install-record.txt --single-version-externally-managed --compile:
    running install
    Package libcgraph was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libcgraph.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libcgraph' found
    Trying pkg-config
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-lDPgM8/pygraphviz/setup.py", line 87, in <module>
        tests_require=['nose>=0.10.1', 'doctest-ignore-unicode>=0.1.0',],
      File "/usr/lib64/python2.7/distutils/core.py", line 151, in setup
        dist.run_commands()
      File "/usr/lib64/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib64/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "setup_commands.py", line 44, in modified_run
        self.include_path, self.library_path = get_graphviz_dirs()
      File "setup_extra.py", line 121, in get_graphviz_dirs
        include_dirs, library_dirs = _pkg_config()
      File "setup_extra.py", line 44, in _pkg_config
        output = S.check_output(['pkg-config', '--libs-only-L', 'libcgraph'])
      File "/usr/lib64/python2.7/subprocess.py", line 573, in check_output
        raise CalledProcessError(retcode, cmd, output=output)
    subprocess.CalledProcessError: Command '['pkg-config', '--libs-only-L', 'libcgraph']' returned non-zero exit status 1
    
    ----------------------------------------
Command "/usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-lDPgM8/pygraphviz/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-Hcz4ba-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-lDPgM8/pygraphviz
```

**Comment by Insoleet** [comment]
You need graphviz installed. Please install it thanks to your package manager ! 

Also, do not use sudo in a virtual env. It will install it in the system instead of the virtualenv.

**Comment by M5oul** [comment]
Ok, I installed graphviz `sudo dnf install graphviz`, then
```bash
(env-sakia) [moul@moulinette sakia]$ pip install pygraphviz
Collecting pygraphviz
  Using cached pygraphviz-1.3.1.tar.gz
Installing collected packages: pygraphviz
  Running setup.py install for pygraphviz
    Complete output from command /home/moul/.pyenv/versions/env-sakia/bin/python3.5 -c "import setuptools, tokenize;__file__='/tmp/pip-build-r1mr5lt0/pygraphviz/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-r2va_ym2-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/moul/.pyenv/versions/env-sakia/include/site/python3.5/pygraphviz:
    running install
    Package libcgraph was not found in the pkg-config search path.
    Perhaps you should add the directory containing `libcgraph.pc'
    to the PKG_CONFIG_PATH environment variable
    No package 'libcgraph' found
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-r1mr5lt0/pygraphviz/setup.py", line 87, in <module>
        tests_require=['nose>=0.10.1', 'doctest-ignore-unicode>=0.1.0',],
      File "/home/moul/.pyenv/versions/3.5.0/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/home/moul/.pyenv/versions/3.5.0/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/home/moul/.pyenv/versions/3.5.0/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/tmp/pip-build-r1mr5lt0/pygraphviz/setup_commands.py", line 44, in modified_run
        self.include_path, self.library_path = get_graphviz_dirs()
      File "/tmp/pip-build-r1mr5lt0/pygraphviz/setup_extra.py", line 121, in get_graphviz_dirs
        include_dirs, library_dirs = _pkg_config()
      File "/tmp/pip-build-r1mr5lt0/pygraphviz/setup_extra.py", line 44, in _pkg_config
        output = S.check_output(['pkg-config', '--libs-only-L', 'libcgraph'])
      File "/home/moul/.pyenv/versions/3.5.0/lib/python3.5/subprocess.py", line 629, in check_output
        **kwargs).stdout
      File "/home/moul/.pyenv/versions/3.5.0/lib/python3.5/subprocess.py", line 711, in run
        output=stdout, stderr=stderr)
    subprocess.CalledProcessError: Command '['pkg-config', '--libs-only-L', 'libcgraph']' returned non-zero exit status 1
    Trying pkg-config
    
    ----------------------------------------
Command "/home/moul/.pyenv/versions/env-sakia/bin/python3.5 -c "import setuptools, tokenize;__file__='/tmp/pip-build-r1mr5lt0/pygraphviz/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-r2va_ym2-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/moul/.pyenv/versions/env-sakia/include/site/python3.5/pygraphviz" failed with error code 1 in /tmp/pip-build-r1mr5lt0/pygraphviz
```

```bash
  File "/home/moul/projects/sakia/build/src/sakia/gui/graphs/explorer_tab.py", line 85, in refresh
    self.graphicsView.scene().update_wot(self.graph.nx_graph, self._current_identity, self.steps_slider.maximum())
  File "/home/moul/projects/sakia/build/src/sakia/gui/views/explorer.py", line 100, in update_wot
    graph_pos = networkx.pygraphviz_layout(nx_graph, prog="twopi")
  File "/home/moul/.pyenv/versions/env-sakia/lib/python3.5/site-packages/networkx/drawing/nx_agraph.py", line 256, in pygraphviz_layout
    'http://pygraphviz.github.io/')
ImportError: ('requires pygraphviz ', 'http://pygraphviz.github.io/')
```

**Comment by Insoleet** [comment]
What does `pkg-config --libs-only-L libcgraph` returns ?

**Comment by M5oul** [comment]
```bash
$ pkg-config --libs-only-L libcgraph
Package libcgraph was not found in the pkg-config search path.
Perhaps you should add the directory containing `libcgraph.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libcgraph' found
```

**Comment by Insoleet** [comment]
Can you please do `ls -l /usr/lib/libcgraph*`? Also, what is your fedora version ?

**Comment by M5oul** [comment]
```bash
 ls -l /usr/lib/libcgraph*
ls: impossible d'accéder à /usr/lib/libcgraph*: Aucun fichier ou dossier de ce type
```
Fedora 23.

**Comment by Insoleet** [comment]
Can you do `echo $PKG_CONFIG_PATH` please ?

**Comment by M5oul** [comment]
Nothing is displayed, the environnement variable doesn't exist in my shell.

**Comment by Insoleet** [comment]
Well, I don't know then. You should report an issue on https://github.com/pygraphviz/pygraphviz/issues . I'll handle the case where pygraphviz is not installed and use default networkx graphs in this situation.

**Comment by Insoleet** [comment]
Ok new build.

I removed the dependency to pygraphviz. I ported the twopi algorithm from graphviz library to python.

Here is the latest render : 
![screenshot_20160108_003947](https://cloud.githubusercontent.com/assets/1170293/12186161/57f17c2a-b5a0-11e5-8e31-01ff5c723f56.png)

One last problem remain, sometimes, nodes are displayed above other nodes. I'm not sure how to handle this.

**Comment by M5oul** [comment]
It's working well on my side. That's awesome!
One thing which could be enhance is that the click on one step don't works. I need to hold then drop on one step.

**Comment by Insoleet** [comment]
Ok, the explorer now has the following animations : 
 - Nodes are expanding from where they were found
 - Current looked up node is shown during research
 - The identity from where we start to do the research stays at the center
 - The research bar is shown like in the wot tab
 - It's possible to reset the explorer tab

I consider the feature finalized. Could you do some quality acceptance checks ? 

Thanks !

**Comment by M5oul** [comment]
That's an awesome feature which works very well, great works :smile_cat:
